si/gb.c

```diff
@@ -210,22 +382,302 @@ void mbc_mbc3_write_clock_data_latch( struct controller *controller, uint16_t ad
+  }
+  
+  // write 0000-1FFF: ram enable
+  for( i=0x00; i<=0x1F; ++i ) {
+    controller->gb_writemem[i] = mbc_mbc5_write_ram_enable;
+  }
+  // write 2000-3FFF: rom bank select
+  for( i=0x20; i<=0x2F; ++i ) {
+    controller->gb_writemem[i] = mbc_mbc5_write_rom_bank_select_lower;
+  }
+  // write 4000-5FFF: ram bank select
+  for( i=0x30; i<=0x3F; ++i ) {
+    controller->gb_writemem[i] = mbc_mbc5_write_rom_bank_select_upper;
+  }
+
+  // write 6000-7FFF: bank mode select
+  for( i=0x40; i<=0x5f; ++i ) {
+    controller->gb_writemem[i] = mbc_mbc5_write_ram_bank_select;
+  }  
+  
+  // read A000-BFFF: read extram
+  // calculate the last address where extram is installed
+  int extram_end = 0xA0 + (controller->cart.extram_size>8192?8192:controller->cart.extram_size)/256;
+  for( i=0xA0; i<extram_end; ++i ) {
+    controller->gb_readmem[i] = mbc_read_extram;
+  }
+  for( i=extram_end; i<=0xBF; ++i ) {
+    controller->gb_readmem[i] = mbc_read_ff;
+  }
+  
+  // write A000-BFFF: write extram
+  for( i=0xA0; i<extram_end; ++i ) {
+    controller->gb_writemem[i] = mbc_write_extram;
+  }
+  for( i=extram_end; i<=0xBF; ++i ) {
+    controller->gb_writemem[i] = mbc_write_dummy;
+  }  
+}
+
+static void mbc_unsupported_install(struct controller *controller) {
+  printf("Transfer Pak ROM type not fully supported. Defaulting to MBC3 behaviour\n");
+  mbc_mbc3_install(controller);
+}
+
+uint8_t gb_read(struct controller *controller, uint16_t address) {
+  uint8_t (*myfunc)(struct controller *,uint16_t) = controller->gb_readmem[address>>8];
+  uint8_t data;
+  if(myfunc != NULL)
+    data = myfunc(controller, address);
+  else
+    data = 0xFF;
+//   printf("gb_read: %04X:%02X\n", address, data);
+  return data;
+}
+
+void gb_write(struct controller *controller, uint16_t address, uint8_t data) {
+  void (*myfunc)(struct controller *,uint16_t,uint8_t) = controller->gb_writemem[address>>8];
+  if(myfunc != NULL)
+    myfunc(controller, address, data);
+//   printf("gb_write: %04X:%02X\n", address, data);
+}
+
+void gb_init(struct controller *controller) {
+  for(int i=0;i<0x100;i++) {
+    controller->gb_readmem [i] = NULL;
+    controller->gb_writemem[i] = NULL;
+  }
+  
+  struct gb_cart *cart = &controller->cart;
+  
+  // set up pak rom
+  cart->cartromsize = controller->tpak_rom.size;
+  cart->cartrom = (uint8_t *)(controller->tpak_rom.ptr);
+  cart->cartrom_bank_zero = cart->cartrom;
+  cart->cartrom_bank_n = cart->cartrom + 0x4000;
+  
+  // set up pak ram
+  if( controller->tpak_save.ptr != NULL ) {
+    cart->extram = (uint8_t *)(controller->tpak_save.ptr);
+    cart->extram_size = controller->tpak_save.size;
+  } else {
+    cart->extram = calloc( 1, 65536 );  // TODO: deallocate this eventually?
+    cart->extram_size = 65536;
+  }   
+  cart->extram_bank = cart->extram;
+
+  uint8_t rom_size_code = cart->cartrom_bank_zero[ROM_SIZE_CODE_ADDRESS];
+  if (rom_size_code <= 8) {
+    cart->cartrom_num_banks = power(2, rom_size_code + 1);
```

Comment:
Why do you need the power method? 2<sup>x</sup> can be calculated in a simpler and faster way with just `1<<x`, so `cart->cartrom_num_banks = 1 << (rom_size_code + 1)` would be a better option here.

Reply:
Thanks for taking a look at it, I have updated the PR with your suggestion.